### ISSUE TITLE
PerlIOUnix_refcnt_dec() croaks in bleadperl since #29065

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8697]
Migrated from [rt.perl.org#41026](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=41026) (status was 'resolved')

Searchable as RT41026$

**Comment by p5pRT** [comment]
### From @steve-m-hay

#### Created by @steve-m-hay



I've found that my Win32\-SharedFileOpen module currently fails a bunch 
of tests using bleadperl&#8203;: the test scripts croak with the error&#8203;:

refcnt\_dec&#8203;: fd 3&#8203;: 0 \<= 0

I've simplified the problem to the attached Foo\-2\.00 module\.

Using perl\-5\.8\.8 \(and bleadperl prior to \#29060\) the test script in this 
Foo module outputs&#8203;:

ok 1
ok 2

Using bleadperl from \#29074 onwards it now croaks with the output&#8203;:

ok 1
refcnt\_dec&#8203;: fd 3&#8203;: 0 \<= 0
refcnt\_dec&#8203;: fd 3&#8203;: 0 \<= 0
refcnt\_dec&#8203;: fd 3&#8203;: 0 \<= 0

The first refcnt\_dec croak\(\) message appears at this point&#8203;:

 >    perl59\.dll\!PerlIOUnix\_refcnt\_dec\(int fd=3\)  Line 2357    C
&nbsp;      perl59\.dll\!PerlIOStdio\_close\(interpreter \* my\_perl=0x00254234\, 
\_PerlIO \* \* f=0x01823f94\)  Line 3115 \+ 0x9 bytes    C
&nbsp;      perl59\.dll\!PerlIO\_\_close\(interpreter \* my\_perl=0x00254234\, \_PerlIO \* 
\* f=0x01823f94\)  Line 1399 \+ 0x10 bytes    C
&nbsp;      perl59\.dll\!Perl\_PerlIO\_close\(interpreter \* my\_perl=0x00254234\, 
\_PerlIO \* \* f=0x01823f94\)  Line 1412 \+ 0xd bytes    C
&nbsp;      perl59\.dll\!Perl\_io\_close\(interpreter \* my\_perl=0x00254234\, io \* 
io=0x0025512c\, char not\_implicit=''\)  Line 977 \+ 0x12 bytes    C
&nbsp;      perl59\.dll\!Perl\_do\_close\(interpreter \* my\_perl=0x00254234\, gv \* 
gv=0x0182c17c\, char not\_implicit=''\)  Line 940 \+ 0x11 bytes    C
&nbsp;      perl59\.dll\!Perl\_pp\_close\(interpreter \* my\_perl=0x00254234\)  Line 600 
\+ 0xf bytes    C
&nbsp;      perl59\.dll\!Perl\_runops\_debug\(interpreter \* my\_perl=0x00254234\)  Line 
1908 \+ 0xf bytes    C
&nbsp;      perl59\.dll\!S\_run\_body\(interpreter \* my\_perl=0x00254234\, long 
oldscope=1\)  Line 2404 \+ 0xf bytes    C
&nbsp;      perl59\.dll\!perl\_run\(interpreter \* my\_perl=0x00254234\)  Line 2324 \+ 
0xd bytes    C
&nbsp;      perl59\.dll\!RunPerl\(int argc=3\, char \* \* argv=0x00274eb0\, char \* \* 
env=0x002731a8\)  Line 248 \+ 0x9 bytes    C\+\+
&nbsp;      perl\.exe\!main\(int argc=3\, char \* \* argv=0x00274eb0\, char \* \* 
env=0x002731a8\)  Line 23 \+ 0x12 bytes    C
&nbsp;      perl\.exe\!\_\_tmainCRTStartup\(\)  Line 586 \+ 0x19 bytes    C
&nbsp;      perl\.exe\!mainCRTStartup\(\)  Line 403    C

The second croak\(\) message comes from within the croak\(\) triggered 
above\, and the third one comes from within perl\_destruct\(\)\.

I'm not sure if the cause of the error is something wrong with my code 
or something wrong with PerlIO\. Either way\, the new croak\(\) message 
added to PerlIOUnix\_refcnt\_dec\(\) by \#29065 is quite possibly just 
revealing a bug that's been there all along\, rather than it or the other 
changes mentioned below having actually broken anything\.

This issue was originally reported on perl5\-porters\, but no solution was 
found\, so I'm perlbugging it&#8203;:

http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2006\-11/msg01062\.html

More background&#8203;:

It works OK at patchlevel 29059\. Change 29060 is probably the culprit\, 
but actually broke the build on Win32 so it is difficult to be sure\. 
Change 29065 fixed the linker error and introduced the actual croak\(\) 
message that I'm getting\, but the "nmake install" step still falls over\. 
That was finally fixed by 29074\, and from that patchlevel onwards my 
module no longer works\.

Change 29060 defined PERLIO\_INIT and PERLIO\_TERM\, and added PERLIO\_TERM 
to PERL\_SYS\_TERM\.

Change 29074 added PERLIO\_INIT to PERL\_SYS\_INIT\, which was apparently 
forgotten in 29060\.

Change 29077 extended the above to other \*ish\.h headers \(only unixish\.h 
and dosish\.h had been addressed until then\)\.

Change 29424 added PERLIO\_TERM to Perl\_win32\_term\(\)\, which was also 
apparently forgotten in 29060\.

<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=high

Site configuration information for perl 5.9.5:

Configured by steveh at Fri Dec  1 10:13:41 2006.

Summary of my perl5 (revision 5 version 9 subversion 5) configuration:
   Platform:
     osname=MSWin32, osvers=5.1, archname=MSWin32-x86-multi-thread
     uname=''
     config_args='undef'
     hint=recommended, useposix=true, d_sigaction=undef
     useithreads=define, usemultiplicity=define
     useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
     use64bitint=undef, use64bitall=undef, uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='cl', ccflags ='-nologo -GF -W3 -Od -MDd -Zi -DDEBUGGING -DWIN32 
-D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT -D_CRT_SECURE_NO_DEPRECATE 
-wd4996  -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO',
     optimize='-Od -MDd -Zi -DDEBUGGING',
     cppflags='-DWIN32'
     ccversion='14.00.50727.42', gccversion='', gccosandvers=''
     intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
     d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
     ivtype='long', ivsize=4, nvtype='double', nvsize=8, 
Off_t='__int64', lseeksize=8
     alignbytes=8, prototype=define
   Linker and Libraries:
     ld='link', ldflags ='-nologo -nodefaultlib -debug 
-libpath:"C:\bleadperl\lib\CORE"  -machine:x86'
     libpth=\lib
     libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib 
  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib 
netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib 
odbc32.lib odbccp32.lib msvcrtd.lib
     perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib 
winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib 
oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib 
version.lib odbc32.lib odbccp32.lib msvcrtd.lib
     libc=msvcrtd.lib, so=dll, useshrplib=yes, libperl=perl59.lib
     gnulibc_version=''
   Dynamic Linking:
     dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
     cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug 
-libpath:"C:\bleadperl\lib\CORE"  -machine:x86'

Locally applied patches:
     DEVEL


@INC for perl 5.9.5:
     C:/bleadperl/lib
     C:/bleadperl/site/lib
     .


Environment for perl 5.9.5:
     HOME (unset)
     LANG (unset)
     LANGUAGE (unset)
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)
     PATH=C:\WINDOWS\system32;C:\WINDOWS;C:\Program Files\Microsoft 
Visual Studio 8\Common7\IDE;C:\Program Files\Microsoft Visual Studio 
8\VC\BIN;C:\Program Files\Microsoft Visual Studio 
8\Common7\Tools;C:\Program Files\Microsoft Visual Studio 
8\Common7\Tools\bin;C:\Program Files\Microsoft Visual Studio 
8\VC\PlatformSDK\bin;C:\Program Files\Microsoft Visual Studio 
8\SDK\v2.0\bin;C:\WINDOWS\Microsoft.NET\Framework\v2.0.50727;C:\Program 
Files\Microsoft Visual Studio 8\VC\VCPackages;C:\bleadperl\bin;C:\cygwin\bin
     PERL_BADLANG (unset)
     SHELL (unset)

-- 


------------------------------------------------
Radan Computational Ltd.

The information contained in this message and any files transmitted with it are confidential and intended for the addressee(s) only. If you have received this message in error or there are any problems, please notify the sender immediately. The unauthorized use, disclosure, copying or alteration of this message is strictly forbidden. Note that any views or opinions presented in this email are solely those of the author and do not necessarily represent those of Radan Computational Ltd. The recipient(s) of this message should check it and any attached files for viruses: Radan Computational will accept no liability for any damage caused by any virus transmitted by this email.

```
</details>

**Comment by p5pRT** [comment]
### From @steve-m-hay

[Foo-2.00.tar.gz](https://rt-archive.perl.org/perl5/Ticket/Attachment/180814/100496/Foo-2.00.tar.gz)

**Comment by p5pRT** [comment]
### From @steve-m-hay

#### Created by @steve-m-hay



I've found that my Win32\-SharedFileOpen module currently fails a bunch 
of tests using bleadperl&#8203;: the test scripts croak with the error&#8203;:

refcnt\_dec&#8203;: fd 3&#8203;: 0 \<= 0

I've simplified the problem to the attached Foo\-2\.00 module\.

Using perl\-5\.8\.8 \(and bleadperl prior to \#29060\) the test script in this 
Foo module outputs&#8203;:

ok 1
ok 2

Using bleadperl from \#29074 onwards it now croaks with the output&#8203;:

ok 1
refcnt\_dec&#8203;: fd 3&#8203;: 0 \<= 0
refcnt\_dec&#8203;: fd 3&#8203;: 0 \<= 0
refcnt\_dec&#8203;: fd 3&#8203;: 0 \<= 0

The first refcnt\_dec croak\(\) message appears at this point&#8203;:

 >    perl59\.dll\!PerlIOUnix\_refcnt\_dec\(int fd=3\)  Line 2357    C
&nbsp;      perl59\.dll\!PerlIOStdio\_close\(interpreter \* my\_perl=0x00254234\, 
\_PerlIO \* \* f=0x01823f94\)  Line 3115 \+ 0x9 bytes    C
&nbsp;      perl59\.dll\!PerlIO\_\_close\(interpreter \* my\_perl=0x00254234\, \_PerlIO \* 
\* f=0x01823f94\)  Line 1399 \+ 0x10 bytes    C
&nbsp;      perl59\.dll\!Perl\_PerlIO\_close\(interpreter \* my\_perl=0x00254234\, 
\_PerlIO \* \* f=0x01823f94\)  Line 1412 \+ 0xd bytes    C
&nbsp;      perl59\.dll\!Perl\_io\_close\(interpreter \* my\_perl=0x00254234\, io \* 
io=0x0025512c\, char not\_implicit=''\)  Line 977 \+ 0x12 bytes    C
&nbsp;      perl59\.dll\!Perl\_do\_close\(interpreter \* my\_perl=0x00254234\, gv \* 
gv=0x0182c17c\, char not\_implicit=''\)  Line 940 \+ 0x11 bytes    C
&nbsp;      perl59\.dll\!Perl\_pp\_close\(interpreter \* my\_perl=0x00254234\)  Line 600 
\+ 0xf bytes    C
&nbsp;      perl59\.dll\!Perl\_runops\_debug\(interpreter \* my\_perl=0x00254234\)  Line 
1908 \+ 0xf bytes    C
&nbsp;      perl59\.dll\!S\_run\_body\(interpreter \* my\_perl=0x00254234\, long 
oldscope=1\)  Line 2404 \+ 0xf bytes    C
&nbsp;      perl59\.dll\!perl\_run\(interpreter \* my\_perl=0x00254234\)  Line 2324 \+ 
0xd bytes    C
&nbsp;      perl59\.dll\!RunPerl\(int argc=3\, char \* \* argv=0x00274eb0\, char \* \* 
env=0x002731a8\)  Line 248 \+ 0x9 bytes    C\+\+
&nbsp;      perl\.exe\!main\(int argc=3\, char \* \* argv=0x00274eb0\, char \* \* 
env=0x002731a8\)  Line 23 \+ 0x12 bytes    C
&nbsp;      perl\.exe\!\_\_tmainCRTStartup\(\)  Line 586 \+ 0x19 bytes    C
&nbsp;      perl\.exe\!mainCRTStartup\(\)  Line 403    C

The second croak\(\) message comes from within the croak\(\) triggered 
above\, and the third one comes from within perl\_destruct\(\)\.

I'm not sure if the cause of the error is something wrong with my code 
or something wrong with PerlIO\. Either way\, the new croak\(\) message 
added to PerlIOUnix\_refcnt\_dec\(\) by \#29065 is quite possibly just 
revealing a bug that's been there all along\, rather than it or the other 
changes mentioned below having actually broken anything\.

This issue was originally reported on perl5\-porters\, but no solution was 
found\, so I'm perlbugging it&#8203;:

http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2006\-11/msg01062\.html

More background&#8203;:

It works OK at patchlevel 29059\. Change 29060 is probably the culprit\, 
but actually broke the build on Win32 so it is difficult to be sure\. 
Change 29065 fixed the linker error and introduced the actual croak\(\) 
message that I'm getting\, but the "nmake install" step still falls over\. 
That was finally fixed by 29074\, and from that patchlevel onwards my 
module no longer works\.

Change 29060 defined PERLIO\_INIT and PERLIO\_TERM\, and added PERLIO\_TERM 
to PERL\_SYS\_TERM\.

Change 29074 added PERLIO\_INIT to PERL\_SYS\_INIT\, which was apparently 
forgotten in 29060\.

Change 29077 extended the above to other \*ish\.h headers \(only unixish\.h 
and dosish\.h had been addressed until then\)\.

Change 29424 added PERLIO\_TERM to Perl\_win32\_term\(\)\, which was also 
apparently forgotten in 29060\.

<details><summary>Perl Info</summary>

```

Flags:
     category=core
     severity=high

Site configuration information for perl 5.9.5:

Configured by steveh at Fri Dec  1 10:13:41 2006.

Summary of my perl5 (revision 5 version 9 subversion 5) configuration:
   Platform:
     osname=MSWin32, osvers=5.1, archname=MSWin32-x86-multi-thread
     uname=''
     config_args='undef'
     hint=recommended, useposix=true, d_sigaction=undef
     useithreads=define, usemultiplicity=define
     useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
     use64bitint=undef, use64bitall=undef, uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='cl', ccflags ='-nologo -GF -W3 -Od -MDd -Zi -DDEBUGGING -DWIN32 
-D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT -D_CRT_SECURE_NO_DEPRECATE 
-wd4996  -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO',
     optimize='-Od -MDd -Zi -DDEBUGGING',
     cppflags='-DWIN32'
     ccversion='14.00.50727.42', gccversion='', gccosandvers=''
     intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
     d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
     ivtype='long', ivsize=4, nvtype='double', nvsize=8, 
Off_t='__int64', lseeksize=8
     alignbytes=8, prototype=define
   Linker and Libraries:
     ld='link', ldflags ='-nologo -nodefaultlib -debug 
-libpath:"C:\bleadperl\lib\CORE"  -machine:x86'
     libpth=\lib
     libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib 
  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib 
netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib 
odbc32.lib odbccp32.lib msvcrtd.lib
     perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib 
winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib 
oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib 
version.lib odbc32.lib odbccp32.lib msvcrtd.lib
     libc=msvcrtd.lib, so=dll, useshrplib=yes, libperl=perl59.lib
     gnulibc_version=''
   Dynamic Linking:
     dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
     cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug 
-libpath:"C:\bleadperl\lib\CORE"  -machine:x86'

Locally applied patches:
     DEVEL


@INC for perl 5.9.5:
     C:/bleadperl/lib
     C:/bleadperl/site/lib
     .


Environment for perl 5.9.5:
     HOME (unset)
     LANG (unset)
     LANGUAGE (unset)
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)
     PATH=C:\WINDOWS\system32;C:\WINDOWS;C:\Program Files\Microsoft 
Visual Studio 8\Common7\IDE;C:\Program Files\Microsoft Visual Studio 
8\VC\BIN;C:\Program Files\Microsoft Visual Studio 
8\Common7\Tools;C:\Program Files\Microsoft Visual Studio 
8\Common7\Tools\bin;C:\Program Files\Microsoft Visual Studio 
8\VC\PlatformSDK\bin;C:\Program Files\Microsoft Visual Studio 
8\SDK\v2.0\bin;C:\WINDOWS\Microsoft.NET\Framework\v2.0.50727;C:\Program 
Files\Microsoft Visual Studio 8\VC\VCPackages;C:\bleadperl\bin;C:\cygwin\bin
     PERL_BADLANG (unset)
     SHELL (unset)

-- 


------------------------------------------------
Radan Computational Ltd.

The information contained in this message and any files transmitted with it are confidential and intended for the addressee(s) only. If you have received this message in error or there are any problems, please notify the sender immediately. The unauthorized use, disclosure, copying or alteration of this message is strictly forbidden. Note that any views or opinions presented in this email are solely those of the author and do not necessarily represent those of Radan Computational Ltd. The recipient(s) of this message should check it and any attached files for viruses: Radan Computational will accept no liability for any damage caused by any virus transmitted by this email.

```
</details>

**Comment by p5pRT** [comment]
### From @steve-m-hay

[Foo-2.00.tar.gz](https://rt-archive.perl.org/perl5/Ticket/Attachment/181317/100565/Foo-2.00.tar.gz)

**Comment by p5pRT** [comment]
### From @smpeters

On Fri Dec 01 21&#8203;:18&#8203;:01 2006\, shay wrote&#8203;:
> This is a bug report for perl from steve\.hay@&#8203;uk\.radan\.com\,
> generated with the help of perlbug 1\.35 running under perl 5\.9\.5\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> I've found that my Win32\-SharedFileOpen module currently fails a bunch 
> of tests using bleadperl&#8203;: the test scripts croak with the error&#8203;:
> 
> refcnt\_dec&#8203;: fd 3&#8203;: 0 \<= 0
> 

This is a duplicate of 41026\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'rejected'

**Comment by p5pRT** [comment]
### From @steve-m-hay

Now fixed in bleadperl by Nicholas with changes \#30610 and \#30633\.

**Comment by p5pRT** [comment]
@steve-m-hay - Status changed from 'new' to 'resolved'
